src/category_theory/endofunctor/algebra.lean

```diff
@@ -194,5 +201,245 @@ lemma str_is_iso (h : limits.is_initial A) : is_iso A.str :=
 
 end initial
 end algebra
+
+/-- A coalgebra of an endofunctor; `str` stands for "structure morphism" -/
+structure coalgebra (F : C ⥤ C) :=
+(V : C)
+(str : V ⟶ F.obj V)
+
+instance [inhabited C] : inhabited (coalgebra (𝟭 C)) := ⟨⟨ default , 𝟙 _ ⟩⟩
+
+namespace coalgebra
+
+variables {F : C ⥤ C} (V : coalgebra F) {V₀ V₁ V₂ : coalgebra F}
+
+/-
+```
+        str
+    V₀ -----> F V₀
+    |          |
+  f |          | F f
+    V          V
+    V₁ -----> F V₁
+        str
+```
+-/
+/-- A morphism between coalgebras of an endofunctor `F` -/
+@[ext] structure hom (V₀ V₁ : coalgebra F) :=
+(f : V₀.1 ⟶ V₁.1)
+(h' : V₀.str ≫ F.map f = f ≫ V₁.str . obviously)
+
+restate_axiom hom.h'
+attribute [simp, reassoc] hom.h
+namespace hom
+
+/-- The identity morphism of an algebra of endofunctor `F` -/
+def id : hom V V := { f := 𝟙 _ }
+
+instance : inhabited (hom V V) := ⟨{ f := 𝟙 _ }⟩
+
+/-- The composition of morphisms between algebras of endofunctor `F` -/
+def comp (f : hom V₀ V₁) (g : hom V₁ V₂) : hom V₀ V₂ := { f := f.1 ≫ g.1 }
+
+end hom
+
+instance (F : C ⥤ C) : category_struct (coalgebra F) :=
+{ hom := hom,
+  id := hom.id,
+  comp := @hom.comp _ _ _ }
+
+@[simp] lemma id_eq_id : coalgebra.hom.id V = 𝟙 V := rfl
+
+@[simp] lemma id_f : (𝟙 _ : V ⟶ V).1 = 𝟙 V.1 := rfl
+
+variables {V₀ V₁ V₂} (f : V₀ ⟶ V₁) (g : V₁ ⟶ V₂)
+
+@[simp] lemma comp_eq_comp : coalgebra.hom.comp f g = f ≫ g := rfl
+
+@[simp] lemma comp_f : (f ≫ g).1 = f.1 ≫ g.1 := rfl
+
+/-- Coalgebras of an endofunctor `F` form a category -/
+instance (F : C ⥤ C) : category (coalgebra F) := {}
+
+/--
+To construct an isomorphism of coalgebras, it suffices to give an isomorphism of the Vs which
+commutes with the structure morphisms.
+-/
+@[simps]
+def iso_mk (h : V₀.1 ≅ V₁.1) (w : V₀.str ≫ F.map h.hom = h.hom ≫ V₁.str ) : V₀ ≅ V₁ :=
+{ hom := { f := h.hom },
+  inv :=
+  { f := h.inv,
+    h' := by { rw [h.eq_inv_comp, ← category.assoc, ←w, category.assoc, ← functor.map_comp],
+               simp only [iso.hom_inv_id, functor.map_id, category.comp_id] } } }
+
+/-- The forgetful functor from the category of coalgebras, forgetting the coalgebraic structure. -/
+@[simps] def forget (F : C ⥤ C) : coalgebra F ⥤ C :=
+{ obj := λ A, A.1,
+  map := λ A B f, f.1 }
+
+/-- A coalgebra morphism with an underlying isomorphism hom in `C` is a coalgebra isomorphism. -/
+lemma iso_of_iso (f : V₀ ⟶ V₁) [is_iso f.1] : is_iso f :=
+⟨⟨{ f := inv f.1,
+    h' := by { rw [is_iso.eq_inv_comp f.1, ← category.assoc, ← f.h, category.assoc], simp } },
+          by tidy⟩⟩
+
+instance forget_reflects_iso : reflects_isomorphisms (forget F) :=
+{ reflects := λ A B, iso_of_iso }
+
+instance forget_faithful : faithful (forget F) := {}
+
+/--
+From a natural transformation `α : F → G` we get a functor from
+coalgebras of `F` to coalgebras of `G`.
+-/
+@[simps]
+def functor_of_nat_trans {F G : C ⥤ C} (α : F ⟶ G) : coalgebra F ⥤ coalgebra G :=
```

Comment:
This was originally past the 100 character mark, one benefit of using `by {...}`
```
def functor_of_nat_trans {F G : C ⥤ C} (α : F ⟶ G) : coalgebra F ⥤ coalgebra G :=
{ obj := λ V,
  { V := V.1,
    str := V.str ≫ α.app V.1 },
  map := λ V₀ V₁ f, { f := f.1,
      h' := by { rw [category.assoc, ← α.naturality, ← category.assoc, f.h, category.assoc] } } }
```